### ISSUE TITLE
Model training is not reproducible: unordered reads + unstable rotation sort inject up to 1.0 AdjEM of churn per retrain

### ISSUE THREAD

**Issue by camvassallo** [camvassallo/cstat#222]
## Summary

None of the three LightGBM trainers in the model tree produce reproducible artifacts. Re-running a trainer against an **unchanged database** yields a different `.onnx`, different early-stopping budgets, and different served predictions. For `roster_impact_model` — the model that produces the served preseason AdjEM projection — the run-to-run churn measures **MAE 0.19 / max 1.02 AdjEM**, with a mean shift of ~0.

This is a correctness-of-process bug rather than a wrong-answer bug: no individual model is "incorrect," but **every retrain injects noise indistinguishable from a real change**, which makes it impossible to verify that a retrain did what it was supposed to do.

Surfaced while working #218 (`roster_adjo` staleness). It is a **blocker for #218's provenance guardrail** — see "Why this blocks #218" below.

## Evidence — the frame is nondeterministic

Two consecutive `build_dataset()` calls in the **same process** against the same DB, sorted identically by `(team_id, season)` before comparison:

```
run1 order a83ce8f3de4e6b80ff7aaad4cea8d8ae   content 60b8639d6f5e35f2f352e8c2437b1127
run2 order a83ce8f3de4e6b80ff7aaad4cea8d8ae   content 715007b8683ec60356f3c4864e5ff950
ORDER identical  : True
CONTENT identical: False
```

Per-column diff of the two runs (4,255 team-seasons):

```
DIFF exp_sr_share      rows=1   e.g. season=2016  0.62328 vs 0.61240
DIFF arch_wizard       rows=1   e.g. season=2023  0.12794 vs 0.12965
DIFF arch_ranger       rows=2   e.g. season=2017  0.13531 vs 0.12893
DIFF arch_barbarian    rows=1   e.g. season=2022  0.27062 vs 0.25923
DIFF arch_monk         rows=1   e.g. season=2022  0.14579 vs 0.15718
```

Note the shape of the result, which pins the mechanism exactly: **only rotation-membership-derived features move.** Every `cam_*` value feature (`cam_wmean`, `cam_sum`, `cam_top1`, `cam_top3_mean`, …) is stable, because tied players have equal cam values so the aggregate is unaffected. Only the class-share and archetype-share features — which are minutes-weighted over the *selected rotation* — change, because a tie decides **who is in the rotation at all**.

## Root cause

Two independent defects that compound, both in `training/train_roster_impact_model.py`:

**1. The rotation sort is unstable and untied.** `aggregate_team_season` (line 286):

```python
g["_rank_key"] = g["campom"].fillna(_NEG)
g = g.sort_values("_rank_key", ascending=False).reset_index(drop=True)
g = g.head(len(CANONICAL_ROTATION_MPG))          # truncate to top 13
g["proj_mpg"] = [CANONICAL_ROTATION_MPG[i] for i in range(len(g))]
```

`pandas.sort_values` defaults to `kind="quicksort"`, which is **not stable**. There is no secondary sort key. So when two players tie on `_rank_key` — including the very common case of **all players with no Torvik coverage tying at the `_NEG` sentinel** — their relative order is arbitrary. `head(13)` then truncates, so a tie straddling the rank-13/14 boundary silently swaps one player out of the rotation for another, and the canonical-MPG ladder reassigns minutes by rank.

**2. The upstream read is unordered.** `PLAYER_QUERY` has no `ORDER BY`, so Postgres may return rows in any order (plan changes, parallel scans, autovacuum, heap churn). That arbitrary order is the input to the unstable sort, and it is *also* directly consumed by LightGBM's row-order-sensitive bagging.

**3. (Related) dependency versions are floor-pinned, not exact.** `training/requirements.txt`:

```
pandas>=2.2
scikit-learn>=1.3
lightgbm>=4.0
numpy>=1.24
```

Even with 1 and 2 fixed, artifacts would not be reproducible across machines or over time.

## Blast radius — measured, not estimated

Retraining `roster_impact_model` on an unchanged DB (the model was restored afterward; nothing was committed):

| | committed (#211) | retrained | delta |
|---|---|---|---|
| `.onnx` md5 | `dce7e6ed33f5…` | `c4e322506e23…` | differs |
| `final_n_estimators` | 259 | 266 | +7 |
| LOSO best-iters (first 4) | 461, 208, 189, 303 | 448, 207, 190, 295 | — |
| in-sample MAE | 3.0913 | 3.0761 | −0.015 |

Served AdjEM prediction delta across all 4,255 team-seasons:

```
mean -0.0018   MAE 0.1923   p95 0.4934   max 1.0203
```

**A team's projected AdjEM can move a full point on a retrain that changed nothing.** The mean shift of ~0 confirms this is symmetric noise, not a systematic improvement.

## Scope — the whole model tree, not just roster_impact

| trainer | `ORDER BY` in reads | row-order-sensitive bagging |
|---|---|---|
| `train_trajectory_model.py` | 0 | `bagging_fraction: 0.7` (line 336) |
| `train_freshman_model.py` | 0 | `bagging_fraction=0.8` (line 231) |
| `train_roster_impact_model.py` | 0 | `bagging_fraction: 0.7` (line 419) |
| `archetypes.py` | 2 | n/a (k-means) |

`archetypes.py` is the only trainer that orders its reads.

This matters more for Layer 1 than Layer 2, because `train_trajectory_model.py` and `train_freshman_model.py` **write the OOF tables** (`trajectory_oof_predictions`, `freshman_oof_predictions:432` TRUNCATE + reload) that the Layer 2 calibrators train on. Nondeterminism there propagates into every downstream model *and* changes the provenance fingerprint, so a no-op retrain of a Layer 1 model would look like a genuine snapshot change.

Note that `lgb_params()` already sets `"seed": 42, "deterministic": True` (lines 427-428) — the intent to be reproducible is there, it's just defeated upstream of LightGBM.

## Why this blocks #218

#218 proposes a provenance stamp so `roster_impact` and `roster_adjo` can be proven to share an OOF snapshot. That guardrail is sound, but it is much weaker while training is nondeterministic:

- **You can't verify a backfilled stamp.** #218's low-blast-radius option was to write the stamp into `roster_impact`'s existing meta without retraining. The natural way to prove that's legitimate — retrain and confirm the artifact is unchanged — does not work today. The retrain produces a different model even when the data is identical.
- **You can't distinguish "retrained correctly" from "retrained differently."** After the #218 retrain, reviewing the diff means asking whether a 0.6-point team move is the honest-OOF fix working or tie-break noise. Currently unanswerable.
- **Every future downstream regen inherits the ambiguity**, which is precisely the class of silent drift #218 exists to prevent.

Landing this first makes the #218 retrain reviewable: with a deterministic frame, the retrain diff is *entirely* attributable to the OOF change, which is the thing being fixed.

## Proposed fix

1. **Deterministic rotation selection.** In `aggregate_team_season`, add an explicit unique tiebreak (e.g. `player_id`) and force a stable sort:
   ```python
   g = g.sort_values(["_rank_key", "player_id"], ascending=[False, True], kind="stable")
   ```
   The tiebreak must be a stable, unique per-player key so the choice is *defined*, not merely repeatable within a session.
2. **`ORDER BY` on every trainer read** — `PLAYER_QUERY`, `TEAM_QUERY`, `OUTBOUND_QUERY`, `INBOUND_QUERY` in `train_roster_impact_model.py`, plus the base queries in `train_trajectory_model.py` and `train_freshman_model.py`. Order on the primary key so the frame is byte-identical across reads.
3. **Exact version pins** in `training/requirements.txt` (`==`) for at least `lightgbm`, `numpy`, `pandas`, `scikit-learn`.
4. **A reproducibility regression test** — build the frame twice and assert the content hash matches, so this can't silently return. Cheap to run in CI (no training needed, just two `build_dataset()` calls). Optionally extend to a full train-twice-and-compare-onnx check, gated behind a slow marker.
5. **Mirror-parity check.** `aggregate_team_season` explicitly "Mirrors `build_roster_impact_features` on the Rust side exactly" (docstring, line 281). The Rust side (`crates/cstat-core/src/roster_impact.rs`) performs the same rank-and-truncate at serve time and needs the *same* tiebreak, or train/serve will disagree on rotation membership for tied players. **This should be verified as part of the fix** — it is a latent train/serve skew independent of reproducibility.

## Acceptance criteria

- [ ] Two consecutive `build_dataset()` calls produce a byte-identical frame (content hash equal), verified by a committed test.
- [ ] Two consecutive full trainer runs on an unchanged DB produce byte-identical `.onnx` for `roster_impact`, `roster_adjo`, `trajectory`, and `freshman`.
- [ ] `training/requirements.txt` pins exact versions.
- [ ] Rust `build_roster_impact_features` uses the same tie-break rule as the Python trainer (or the divergence is documented as immaterial with evidence).
- [ ] No model artifact is retrained/committed in this PR — it is pure infrastructure, so the diff is reviewable on its own. The first deterministic artifacts land with #218.

## Out of scope

- Retraining any model (that is #218 and the downstream program).
- Changing model hyperparameters or features. The tie-break changes rotation membership for a handful of tied players, which is a *behavior* change to the frame; it should be called out in the #218 retrain diff, not smuggled in here as a silent model update.

## Notes

- Reproducing the measurement: restore the committed model with `git show HEAD:training/models/roster_impact_model.onnx`, retrain, and compare ONNX predictions over `build_dataset()`'s frame.
- `training/models/roster_impact_loso/*` is gitignored and regenerated by `train_roster_impact_model.py`. It is consumed by `cstat-ingest projections-backtest` (`crates/cstat-ingest/src/projections_backtest.rs:329`), whose JSON dumps are in turn consumed by `training/compute_cae.py:69`. These artifacts must be regenerated together; a stale mix is currently undetectable.

**Comment by camvassallo** [comment]
## Two findings from implementing this

**1. A third defect: the ONNX export was not byte-stable either.**

After fixing the rotation tie-break and the frame ordering, the *fit* was fully deterministic — identical `final_n_estimators` and identical per-fold LOSO early-stopping vectors across runs — but the `.onnx` bytes still differed. `onnxmltools.convert_lightgbm` names the graph with a fresh UUID on every conversion:

```
graph.name: 77a81725546c4599850faa3a5b5cd823  vs  1f267ec431c3440492d06fb702af035c
predictions identical: True   max|diff| = 0.0
```

Functionally identical, byte-different. That defeats the point of the other two fixes — the reason to make training deterministic is to be able to read a model diff and conclude "nothing changed". Fixed by naming the graph after the artifact. Worth knowing this applies to `export_onnx.py` and `train_roster_model.py` too, which are outside this issue's scope (the game models have no edge to the roster tree).

**2. `ORDER BY` on the natural key is not sufficient — noted because it looks like it should be.**

`player_season_stats` is unique on `(player_id, team_id, season)` and `recruits` can hold more than one row per player, so several joins fan out and the natural key does not determine a row. Both Layer 1 queries stayed unstable after ordering on their key, and again after adding the team ids. The fix that is actually provable is sorting on the full column tuple (`db.canonical_frame_order`): two rows equal across every selected column are byte-identical, so their order cannot affect frame content regardless of what the fan-out does.

## Out of scope, but surfaced here

The fan-out above means the Layer 1 training frames carry **exact duplicate rows** — 274 in the trajectory frame, 35 in the freshman frame — which double-weights multi-team player-seasons in training. A player who appears on two teams in one season contributes two identical rows.

This is a modeling question, not a determinism one: deduplicating would change the fitted models, so it does not belong in a pure-infrastructure PR. Recording it here rather than opening a separate issue until someone decides whether the double-weighting is wrong. Arguments both ways — it is plausibly a bug (the player is one observation), but it is also a mild upweighting of transfer/multi-team players who are arguably harder to project and worth more weight. Wants a LOSO A/B before changing.

**Comment by camvassallo** [comment]
Heads up that this regressed, and its own guard has been failing on `main` — found while running the training guards against a live database for #311/#319.

```
$ cd training && ./.venv/bin/python test_frame_determinism.py
1 failure(s):
  - test_roster_impact_frame_is_deterministic: build_dataset() returned two
    different frames from the same database. The read is not deterministic (issue #222).
```

Confirmed against `origin/main`'s `train_roster_impact_model.py`, not just the #319 branch.

**Why CI never caught it.** The `Training Guards` job points `DATABASE_URL` at a deliberately unreachable host, so `_db_reachable()` returns false and every frame check *skips*. That is the documented and correct choice — the alternative, the test job's empty Postgres, would fail on missing tables rather than skip — but it does mean the four database-backed determinism checks have no CI coverage at all. Only a local run with a populated database sees them.

**Cause.** Exactly the failure mode this issue is about, in a query it did not reach: `OUTBOUND_QUERY`'s

```sql
COALESCE(SUM(COALESCE(tps.cam_gbpm_v3_psos, 0)), 0)::float8 AS outbound_cam_v3_sum
```

Floating-point addition is not associative, so an unordered `SUM` folds each group in whatever order the plan emits. Two reads of the same database differed on `outbound_cam_v3_sum` for **245 of 4,255 team-seasons**, purely in the last bits:

```
a = 7.090762969719948   b = 7.090762969719947
a = -2.5025741984017293 b = -2.50257419840173
```

Which is precisely the churn this issue was opened about — it lands on LightGBM split thresholds, and it makes a retrain's provenance fingerprint move so a real model diff can't be told from summation noise.

**Fixed in #319** by pinning the fold order inside the aggregate, `SUM(... ORDER BY p_base.id)`, on both `OUTBOUND_QUERY` and `INBOUND_QUERY` — the inbound one did not happen to differ across the two reads the guard takes, which is the reason to pin it too: it is stable per plan, not across plans. `test_frame_determinism.py` now reports 5/5.

It rode along there rather than getting its own PR because #319 was already rewriting those two aggregates for the Torvik fan-out, and because the retrain #319 requires is not reproducible until this is fixed — a retrain on a non-deterministic frame cannot be attributed.

Worth reopening, or opening a follow-up, for the coverage gap itself: nothing stops the next regression either. Options are a CI job with a seeded database, or making `_db_reachable()`-skipped checks a visible non-zero-exit condition in a nightly rather than a silent skip.
